docs/pallas/tpu/pipelining.md

```diff
@@ -0,0 +1,397 @@
+  )(x, y)
+  return z
+
+
+x, y = jnp.ones((512, 512)), jnp.ones((512, 512))
+add_matrices(x, y)
+```
+
++++ {"id": "HMENNLy8okCL"}
+
+We've written two functions: `add_matrices_kernel` and `add_matrices`.
+
+`add_matrices_kernel` operates using `Ref`s that live in VMEM. Loading from a VMEM `Ref` produces a value that lives in VREGs. Values in VREGs behave like `jax.Array`s in that we can use `jnp` and `jax.lax` operations on then to produce new values that live in VREGs. When we produce the values we'd like to return, we store them in the output VMEM `Ref`.
+
+The `add_matrices` function acts on `jax.Array`s and returns a `jax.Array`. Inside it, we pass `x` and `y` into `pallas_call`. `pallas_call` is responsible for copying `x` and `y` into VMEM and for allocating the VMEM buffers that the kernel operates on (including allocating `z_vmem_ref`, the output VMEM buffer). After the kernel function is finished running, `pallas_call` will also copy the value in `z_vmem_ref` to HBM, resulting in an output `jax.Array`.
+
++++ {"id": "5kWr-1tKpYro"}
+
+## Constraints of using VMEM/SMEM
+
+Pallas exposes access to lower level memory spaces like VMEM and SMEM but writing kernels utilizing them adds some considerations.
+
+1. Memory capacity. VMEM and SMEM are *small*! VMEM on v4 TPUs is only 16MiB and SMEM ranges in the tens to hundreds of KiB. If our arrays are too big, we won't even be able to fit them into VMEM at all. For reference, a `f32[2048, 2048]` array is 16MiB, so our above kernel won't scale beyond moderately sized arrays.
+
+2. Memory bandwidth. Copying to/from HBM and VMEM takes a long time, at least compared to most compute instructions. The `add_matrices` function above will likely spend more time copying between HBM and VMEM than actually performing the addition itself.
+
+With these two constraints in mind, we'll have to rethink our strategy for getting performance of our TPUs.
+
++++ {"id": "_NTqvlbetB3P"}
+
+## Primer: Pipelining
+
+Pipelining our computation offers a way of dealing with both the memory capacity and bandwidth constraints in one fell swoop. What do we mean by pipelining?
+
+The goal is: *concurrently* copy to/from HBM and VMEM *while* utilizing our compute units. However, the way we expressed our program above is that we copy *all* of `x` and `y` before we start doing any compute with them. This creates a dependence between the copy and the compute.
+
+However, if we can chunk up our computation into several subcomputations (e.g. when we add two matrices, we can express that as addition of "blocks" of the original matrices together), we can now overlap the copies of one of those subcomputations with the compute of the other. Let's walk through a simple example:
+
+Let's say we split our arrays `x` and `y` into `x1, x2` and `y1, y2` (for example, split along the leading axis, resulting in two `(256, 512)` arrays for each input. We can now execute the following pipelined computation.
```

Comment:
Maybe we can make a pair of timeline sketches illustrating how things look with and without pipelining for this example? (Or even better than a drawing would be if we could 'take a photo' of real timings, but I'm not sure if that's feasible.)

Reply:
Anselm pointed me to how to make animations of this, so I'll try that and add it as a follow up!